File: 2024/src/day01.rs
use std::time::Instant;

const LIST_LEN: usize = 1000;

pub(crate) fn solve() {
    let input = include_str!("day01_input.txt");
    let (mut left, mut right) = parse_lists(input);
    left.sort();
    right.sort();

    let instant_part1 = Instant::now();
    let mut sum: i64 = 0;
    for i in 0..LIST_LEN {
        sum += (left[i] - right[i]).abs();
    }
    println!("Part 1 took: {:?}", instant_part1.elapsed());

    let instant_part2 = Instant::now();
    let mut similarity: i64 = 0;
    for i in left.iter() {
        for j in right.iter() {
            if i == j {
                similarity += i;
            } else if i < j {
                break;
            }
        }
    }
    println!("Part 2 took: {:?}", instant_part2.elapsed());

    println!("Day 1 Part 1: {}", sum);
    println!("Day 1 Part 2: {}", similarity);
}

fn parse_lists(input: &str) -> (Vec<i64>, Vec<i64>) {
    let instant = Instant::now();
    let mut left = Vec::with_capacity(LIST_LEN);
    let mut right = Vec::with_capacity(LIST_LEN);

    for line in input.lines() {
        let mut split_iter = line.split_ascii_whitespace();
        left.push(split_iter.next().unwrap().parse::<i64>().unwrap());
        right.push(split_iter.next().unwrap().parse::<i64>().unwrap());
    }

    println!("Parsing lists took: {:?}", instant.elapsed());

    assert!(left.len() == LIST_LEN);
    assert!(left.len() == right.len());

    (left, right)
}

File: 2024/python/src/day01.py
import time
from pathlib import Path

input_file = Path(__file__).parent / "day01_input.txt"
input = input_file.read_text()

t1 = time.perf_counter_ns()

left: list[int] = []
right: list[int] = []

for line in input.splitlines():
    left.append(int(line[:5]))
    right.append(int(line[8:]))

left.sort()
right.sort()

sum = 0
for left_item, right_item in zip(left, right, strict=True):
    sum += abs(left_item - right_item)
print(f"Part 1: {sum}")

similarity = 0
lookup = [0] * 100_000
for i in right:
    lookup[i] += i
for i in left:
    similarity += lookup[i]
print(f"Part 2: {similarity}")

t2 = time.perf_counter_ns()
print(f"Time: {(t2 - t1) / 1_000_000:.3f}ms")
